### ISSUE TITLE
fontforge Creates invalid .otf on i386 and hppa, causing FTBFS 3270font

### ISSUE THREAD

**Issue by copyninja** [fontforge/fontforge#3059]
Hi,

I'm forwarding [Debian bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=831425) here regarding a failure while trying to build 3270font from source. Though the bug is reported on older version of fontforge the same issue still happens on latest version of fontforge as seen [here](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=860649).

**Comment by jtanx** [comment]
Some relevant links
http://aws-logs.debian.net/2017/04/18/3270font_1.2.20-1_testing-i386.log
https://github.com/rbanffy/3270font

FWIW it seems to generate fine on Windows with a 32 bit build. I don't have very good internet right now, so I can't test on Debian/i386 :(

**Comment by copyninja** [comment]
@jtanx any updates?.. If you need help like trying to build that font on hppa let me know.

**Comment by jtanx** [comment]
@copyninja unfortunately no - I'd like to test on i386 but I don't have the bandwidth to download it .

**Comment by jtanx** [comment]
Interesting... I think I've been able to reproduce this issue on Debian Jessie/i386

**Comment by jtanx** [comment]
After some further digging around, I can narrow it down to something being different depending on compiler optimisation for the `comparespline` function in `splineorder2.c` - when the optimisation level is at `O2` or higher, the bug is apparent. Otherwise, it's fine.

**Comment by copyninja** [comment]
@jtanx Thanks for  hints In Debian the CFLAGS involves -O2 and I will see if I can remove that flag after discussion with fonts team.

**Comment by jtanx** [comment]
@copyninja I think disabling O2 optimisation globally is not ideal... If possible, disabling just for that file or function would be better. 

In:

```c
static int comparespline(Spline *ps, Spline *ttf, real tmin, real tmax, real err) {
    int dim=0, other;
    real dx, dy, ddim, dt, t;
    real d, o;
```

Changing this to:

```c
static int comparespline(Spline *ps, Spline *ttf, real tmin, real tmax, real err) {
    int dim=0, other;
    real dx, dy, ddim, dt, t;
    volatile real d, o;
```

Seems to be enough to prevent it optimising too far.

-----

That said, I'm still trying to see _why_ optimisation is causing the difference :S

**Comment by jtanx** [comment]
Ahh, I think I've figured out why this is happening. I'd say it's a compiler/gcc bug.

For:

```c
for ( t=.1; t<1; t+= .1 ) {
	d = (ttf->splines[0].b*t+ttf->splines[0].c)*t+ttf->splines[0].d;
	o = (ttf->splines[1].b*t+ttf->splines[1].c)*t+ttf->splines[1].d;
	if ( d<bb.minx || d>bb.maxx || o<bb.miny || o>bb.maxy )
return( false );
    }
```

I was noticing that under optimisation, gcc was turning the multiplication instructions into [single precision multiplications](https://gist.github.com/jtanx/b23d1fe2a7473d110d9d368f9cd52d3d#file-comparespline-o2-txt-L120), whereas in the less optimised versions, it was doing a [double precision multiplication](https://gist.github.com/jtanx/b23d1fe2a7473d110d9d368f9cd52d3d#file-comparespline-o1-txt-L154), as expected.

I'm guessing that gcc is seeing the literal values of `.1` and `1`, and for some reason, optimising this away into a floating point multiplication when it should not. Changing the code to read:

```c
for ( t=.1L; t<1L; t+= .1L ) {
...
```

Fixes it, but FF has configurable precision for its real type, so I'm not sure if this would break for builds of `real=float`.

**Comment by copyninja** [comment]
@jtanx Discussed with Paul Wise of our team and he suggested use of `creduce` on the above case and report it to the GCC upstream. I'm not familiar with creduce but if I get time I will try to use it to see how it can be done.

**Comment by jtanx** [comment]
Mmm possibly, and hope that the compiler doesn't optimise that away too. Then you have to check everywhere else that literals are used for the same possible problem...

**Comment by jtanx** [comment]
Hah, I can't believe I debugged this, forgot about it, then ran into the same problem down the track (#4290).

Anyway this should be fixed with 0f049ccfd4dacea994d060c41cbb57b34766d1d1. Or alternatively by passing `-msse2 -mfpmath=sse` although I guess that only works on i386.
